share/completions/clj.fish

```diff
@@ -0,0 +1,79 @@
+  (->> (clojure.edn/read-string (slurp path))
+       :aliases
+       keys
+       (map name)
+       sort))
+
+(defn installed-tools
+  []
+  (as-> config $
+    (:config-dir $)
+    (str $ "/tools/")
+    (fs/match $ "regex:.*\\\\.edn")
+    (map fs/file-name $)
+    (map #(fs/strip-ext % {:ext "edn"}) $)))
+
+(defn aliases []
+  (->> (configs)
+       (mapcat aliases-from-config)
+       dedupe))
+
+(run! println
+      (case (first *command-line-args*)
+        "tools" (installed-tools)
+        (aliases)))'
+
+function __fish_clj_aliases
+    which bb >/dev/null 2>&1; or return 1
+    bb (echo $bb_helper | psub)
+end
+
+function __fish_clj_tools
+    which bb >/dev/null 2>&1; or return 1
+    bb (echo $bb_helper | psub) tools
+end
+
+complete -c clj -s X -x -r -k -a "(__fish_complete_list : __fish_clj_aliases)" -d "Use concatenated aliases to modify classpath or supply exec fn/args"
+complete -c clj -s A -x -r -k -a "(__fish_complete_list : __fish_clj_aliases)" -d "Use concatenated aliases to modify classpath"
+complete -c clj -s M -x -r -k -a "(__fish_complete_list : __fish_clj_aliases)" -d "Use concatenated aliases to modify classpath or supply main opts"
+complete -c clj -s T -x -r -k -a "(__fish_complete_list : __fish_clj_tools)" -d "Invoke tool by name or via aliases ala -X"
+
+complete -c clj -f -o Sdeps -r -d "Deps data to use as the last deps file to be merged"
+complete -c clj -f -o Spath -d "Compute classpath and echo to stdout only"
+complete -c clj -f -o Spom -d "Generate (or update) pom.xml with deps and paths"
+complete -c clj -f -o Stree -d "Print dependency tree"
+complete -c clj -f -o Scp -r -d "Do NOT compute or cache classpath, use this one instead"
+complete -c clj -f -o Srepro -d "Ignore the ~/.clojure/deps.edn config file"
+complete -c clj -f -o Sforce -d "Force recomputation of the classpath (don\'t use the cache)"
+complete -c clj -f -o Sverbose -d "Print important path info to console"
+complete -c clj -f -o Sdescribe -d "Print environment and command parsing info as data"
+complete -c clj -f -o Sthreads -d "Set specific number of download threads"
+complete -c clj -f -o Strace -d "Write a trace.edn file that traces deps expansion"
+
+complete -c clj -k -f -s h -l help -d "Show help"
+complete -c clj -f -s P -d "Prepare deps - download libs, cache classpath, but don\'t exec"
+complete -c clj -s J -d "Pass opt through in java_opts, ex: -J-Xmx512m"
+complete -c clj -s i -l init -r -d "Load a file or resource"
+complete -c clj -f -s e -l eval -r -d "Eval exprs in string; print non-nil values"
+complete -c clj -f -l report -d "Report uncaught exception to \"file\" (default), \"stderr\", or \"none\""
+complete -c clj -f -o version -l version -d "Print the version to stdout and exit"
+complete -c clj -f -s m -l main -d "Call the -main function from namespace w/args"
+complete -c clj -f -s r -l repl -d "Run a repl"
+complete -c clj -r -d "Run a script from a file or resource"
+
+complete -c clojure --wraps clj
```

Comment:
This won't be loaded at the right time.

Fish finds completion files by command name, so if you try to complete "clojure" it will look for "clojure.fish".

That means once you have tried completing "clj" in the current session, the wrapper will be set up. But if you start a new shell and try to complete "clojure" before "clj" it won't work.

You'll have to add a file called "clojure.fish" that includes just this single line.

However: That's only if "clojure" is a common name set *by upstream* - if it's typically *installed* as both "clojure" and "clj".

If users usually set this up themselves, e.g. by using `alias clojure=clj`, then please just drop this. (`alias` would set up the wrapping)

Reply:
Both are typically installed and are widely used. `clj` is just `clojure` with a `rlwrap` wrapper for use in the repl. I'll make a separate clojure.fish file.

Thank you for the pointers. I'll fix the issues you've mentioned.

Clojure's command line is a bit odd. For the execution options -A, -M, -X and -T, it will take a colon separated list of aliases without any spaces between the option and the list (like `clj -Mdev:debug:build`). `__fish_complete_list` handles this nicely, but it would be nice if the description changed when making list suggestions, so that it first shows

```
❯ clj -<tab>
-A                      (Use concatenated aliases to modify classpath)  
...
```

but once the user has typed -A and hits tab it would show something like

```
❯ clj -Ab<tab>
-Abuild                 (Alias)  
-Abenchmark             (Alias)  
...
```

I was not able to figure out how to do this. It is possible?

Reply:
If the completions for `clj` and `clojure` are identical or close to it, you can always put them in a function that takes the command name as a parameter, and invoke that from both `clj.fish` and `clojure.fish`. There are other workarounds such as defining `clj` as `function clj --wraps clojure; command clj; end;` but that's the kind of thing you would do on your pc and not as part of the base fish installation.